content/blog/2024/testing-fastapi.md

```diff
@@ -0,0 +1,355 @@
+
+    # Rollback the overall transaction, restoring the state before the test ran
+    session.close()
+    transaction.rollback()
+    connection.close()
+```
+
+This fixture has the `autouse` flag, which means it will be automatically used
+by all tests. It creates a new connection to the database using the `engine`
+from the previous fixture and ensures all changes are rolled back after the test
+is complete. The actual override of our application happens in this line:
+
+```python
+app.dependency_overrides[get_db] = lambda: session
+```
+
+This tells FastAPI to use the `session` object instead of the real `get_db`
+function when resolving the `db` parameter in our endpoint. Now all endpoints
+that use `get_db` will use our automatically rolled-back session instead of the
+real database, giving us predictable and isolated tests.
+
+Additionally, we can also use the `db` fixture to assert the application state
+after making requests to our endpoints:
+
+```python
+# File: test_api.py
+
+from app.models import Item
+
+def test_add_item(db):
+    response = client.post("/add-item", json={"name": "Item 1"})
+    assert response.status_code == 200
+    item = db.scalars(select(Item)).one()
+    assert item.name == "Item 1"
+```
+
+The `db` fixture allows to make queries in the context of the test transaction
+and is automatically rolled back after the test is complete.
+
+## Testing endpoints that require authentication
+
+We can easily create a fixture to get instances of the test client for our
+tests:
+
+```python
+# File: conftest.py
+
+from fastapi.testclient import TestClient
+from app.main import app
+
+@pytest.fixture
+def anon_client():
+    return TestClient(app)
+```
+
+We can then use this fixture in our tests to make anonymous requests to our
+application:
+
+```python
+# File: test_api.py
+
+def test_root(anon_client):
+    response = anon_client.get("/")
+    assert response.status_code == 200
+```
+
+But what if we need to test endpoints that require authentication? We can create
+a new fixture that creates a user and an access token and injects them into the
+client:
+
+```python
+# File: conftest.py
+
+from fastapi.testclient import TestClient
+from app.main import app
+
+# These functions are defined somewhere in the application code
+from app.auth import create_access_token, create_user
+
+@pytest.fixture
+def client(db):
+    user = create_user(db)
+    token = create_access_token(user)
+    client = TestClient(app)
+    client.user = user
+    client.headers["Authorization"] = f"Bearer {token}"
+    return client
+```
+
+_`client.user = user` is a convenience to access the user instance in tests,
+nothing else._
+
+Now we can use the `client` fixture in our tests to make requests to our
+application with a user that is authenticated. Assuming our application checks
+for tokens in the `Authorization` header, we can now test our authenticated
+endpoints:
+
+```python
+# File: test_api.py
+
+def test_protected_endpoint(client):
```

Comment:
Mostly out of curiosity, am I understanding this correctly? You declare a def in `conftest.py` called `client`. Then, in your test file, you name your parameter the same thing as the def, and then the value gets passed in? In other words, if you wrote `def test_protected_endpoint(anon_client)` here, it would pass the anon client? Or if you wrote `def test_protected_endpoint(db)` it would insert the `db` you added in conftest.py?